tests/generate_template_reference_data.py

```diff
@@ -143,16 +145,24 @@ def _store_bundled(self, operations):
         return bid
 
 
-def get_masked_flowproject(p, **kwargs):
+@contextlib.contextmanager
+def get_masked_flowproject(p, environment=None):
     """Mock environment-dependent attributes and functions. Need to mock
     sys.executable before the FlowProject is instantiated, and then modify the
     root_directory and project_dir elements after creation."""
-    sys.executable = "/usr/local/bin/python"
-    fp = TestProject.get_project(root=p.root_directory(), **kwargs)
-    fp._entrypoint.setdefault("path", "generate_template_reference_data.py")
-    fp.root_directory = lambda: PROJECT_DIRECTORY
-    fp.config.project_dir = PROJECT_DIRECTORY
-    return fp
+    try:
+        old_executable = sys.executable
+        sys.executable = "/usr/local/bin/python"
+        fp = TestProject.get_project(root=p.root_directory())
+        if environment is not None:
+            fp._environment = environment
+        fp._entrypoint.setdefault("path", "generate_template_reference_data.py")
+        fp._mock = True
+        fp._mock_root_directory = lambda: PROJECT_DIRECTORY
+        fp.config.project_dir = PROJECT_DIRECTORY
```

Comment:
These appear superfluous to me.
```suggestion
        fp.config.project_dir = PROJECT_DIRECTORY
```

Reply:
Never mind, I see where these are used now. I would rather set `fp.root_directory = lambda: PROJECT_DIRECTORY` here than include code in the main code base that exists just to facilitate this test. Is there a reason that wouldn't work?

Reply:
Yes, that won't work because we have to mock the root directory used _for operation id generation_ separately from the real root directory that contains the jobs. We need consistently generated ids (using the mocked value) but we still need the root directory to be correct or else we break signac (maybe instantiating jobs fails? Or the config is inaccessible? Or it can't pickle/unpickle? Something like that, but I forgot the exact reason).

Reply:
Got it, you're right that this won't work in that case. How about the following context manager instead:

```
@contextmanager
def wrap_id_generation(fp):
    """Create a context where id generation will use a mocked root directory."""
    orig_generate_id = fp._generate_id 
    orig_root_directory = fp.root_directory
    mocked_root_directory = lambda: PROJECT_DIRECTORY

    def wrapped_generate_id(self, aggregate, operation_name=None):
        """Mock the root directory used for id generation to ensure consistent ids."""
        # This mocking has to happen within this method to avoid affecting other methods called during the test.
        self.root_directory = mocked_root_directory
        _generate_id(aggregate, operation_name)
        self.root_directory = orig_root_directory
    
    fp._generate_id = wrapped_generate_id 
    yield
    fp._generate_id = orig_generate_id
```
We don't execute anything in parallel here, so I think this should be safe. It's a few more lines of code than the current solution, but I would really like to avoid injected this code into flow/project.py if at all possible.

Reply:
You're right, this would be much better. I'll adapt this.